assembly/src/assembler/mod.rs

```diff
@@ -595,86 +599,154 @@ impl Assembler {
 
         for op in body {
             match op {
                 Op::Inst(inst) => {
-                    if let Some(mast_node_id) = self.compile_instruction(
+                    if let Some(node_id) = self.compile_instruction(
                         inst,
                         &mut basic_block_builder,
                         proc_ctx,
                         mast_forest_builder,
                     )? {
-                        if let Some(basic_block_id) =
-                            basic_block_builder.make_basic_block(mast_forest_builder)?
-                        {
-                            node_ids.push(basic_block_id);
+                        match basic_block_builder.make_basic_block(mast_forest_builder)? {
+                            BasicBlockOrDecorators::BasicBlock(basic_block_id) => {
+                                body_node_ids.push(basic_block_id);
+                            },
+                            BasicBlockOrDecorators::Decorators(decorator_ids) => {
+                                mast_forest_builder.set_before_enter(node_id, decorator_ids);
+                            },
+                            BasicBlockOrDecorators::Nothing => (),
                         }
 
-                        node_ids.push(mast_node_id);
+                        body_node_ids.push(node_id);
                     }
                 },
 
                 Op::If { then_blk, else_blk, .. } => {
-                    if let Some(basic_block_id) =
-                        basic_block_builder.make_basic_block(mast_forest_builder)?
-                    {
-                        node_ids.push(basic_block_id);
-                    }
+                    let maybe_pre_decorators: Option<Vec<DecoratorId>> = match basic_block_builder
+                        .make_basic_block(
+                        mast_forest_builder,
+                    )? {
+                        BasicBlockOrDecorators::BasicBlock(basic_block_id) => {
+                            body_node_ids.push(basic_block_id);
+                            None
+                        },
+                        BasicBlockOrDecorators::Decorators(decorator_ids) => Some(decorator_ids),
+                        BasicBlockOrDecorators::Nothing => None,
+                    };
 
                     let then_blk =
                         self.compile_body(then_blk.iter(), proc_ctx, None, mast_forest_builder)?;
                     let else_blk =
                         self.compile_body(else_blk.iter(), proc_ctx, None, mast_forest_builder)?;
 
                     let split_node_id = mast_forest_builder.ensure_split(then_blk, else_blk)?;
-                    node_ids.push(split_node_id);
+                    if let Some(pre_decorator_ids) = maybe_pre_decorators {
+                        mast_forest_builder.set_before_enter(split_node_id, pre_decorator_ids)
+                    }
+
+                    body_node_ids.push(split_node_id);
```

Comment:
Done.

Although I was not able to apply the same simplification to `BasicBlockBuilder::try_into_basic_block()`, since the builder is consumed in the process, and so we need to also return the decorators from that call (if any).